src/future/shared.rs

```diff
@@ -51,85 +70,77 @@ impl<F> Future for Shared<F>
     type Error = SharedError<F::Error>;
 
     fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
-        let mut inner = self.inner.lock().unwrap();
-        let result = match *inner {
-            State::Waiting(ref mut future, _) => Some(future.poll()),
-            State::Done(_) => None,
-        };
-        let new_state = match result {
-            Some(Ok(Async::NotReady)) => None,
-            Some(Ok(Async::Ready(e))) => Some(State::Done(Ok(Arc::new(e)))),
-            Some(Err(e)) => Some(State::Done(Err(Arc::new(e)))),
-            None => None,
-        };
-        let tasks_to_wake = match new_state {
-            Some(new) => {
-                match mem::replace(&mut *inner, new) {
-                    State::Waiting(_, tasks) => tasks,
-                    State::Done(_) => panic!(),
+        let state = self.inner.state.read().unwrap();
+        let event = match *state {
+            State::Waiting(ref unparker) => {
+                if unparker.original_future_needs_poll.swap(false, Ordering::SeqCst) {
+                    task::UnparkEvent::new(unparker.clone(), 0)
+                } else {
+                    unparker.insert(self.id, task::park());
+                    return Ok(Async::NotReady)
+                }
+            }
+            State::Done(ref r) => {
+                match *r {
+                    Ok(ref v) => return Ok(SharedItem { item: v.clone() }.into()),
+                    Err(ref e) => return Err(SharedError { error: e.clone() }.into()),
                 }
             }
-            None => Vec::new(),
         };
-
-        let ret = match *inner {
-            State::Waiting(_, ref mut tasks) => {
-                tasks.push(task::park());
-                Ok(Async::NotReady)
+        drop(state);
+
+        let mut original_future = self.inner.original_future.lock().unwrap();
+        let done_val = match *original_future {
+            Some(ref mut future) => {
+                match task::with_unpark_event(event, || future.poll()) {
+                    Ok(Async::NotReady) => return Ok(Async::NotReady),
+                    Ok(Async::Ready(v)) => Ok(Arc::new(v)),
+                    Err(e) => Err(Arc::new(e)),
+                }
             }
-            State::Done(Ok(ref e)) => Ok(SharedItem { item: e.clone() }.into()),
-            State::Done(Err(ref e)) => Err(SharedError { error: e.clone() }.into()),
+            None => unreachable!(),
```

Comment:
Hm. On second thought, the new branch that I added does not only apply to the recursive case. If a `poll()` on a separate thread fails to grab the `original_future` mutex, it could now get into a busy wait loop of repeatedly calling `task::park().park()`.